NetworkProcess does not always terminate properly on macOS

bnham opened the pull request:
#### 861148b958c311059840a4b04d00750d461d32e6
<pre>
NetworkProcess does not always terminate properly on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=285947">https://bugs.webkit.org/show_bug.cgi?id=285947</a>
<a href="https://rdar.apple.com/142838090">rdar://142838090</a>

Reviewed by Per Arne Vollan.

NetworkProcess does not always terminate on macOS. The reason for this is that NetworkProcess can
suspend in the middle of termination:

1. The NetworkProcessProxy destructor drops &quot;m_backgroundActivityToPreventSuspension&quot;, which allows
NetworkProcess to suspend.
2. Then the AuxiliaryProcessProxy destructor runs and starts termination by invalidating via
IPC::Connection::invalidate.
3. The remote side of the connection invalidation in NetworkProcess (NetworkProcess::didClose) might
never run if the process is already suspended, which then leaves the process in a permanently
suspended state.

To fix this, as part of AuxiliaryProcessProxy destruction, we now start a watchdog timer which
allows the process to keep running while termination is active, and we kill the process when the
timer expires. This used to be part of XPCConnectionTerminationWatchdog, but I removed that class
since it wasn&apos;t really doing much anymore after this refactoring.

The main refactoring was making it so that process assertions can be created without an
AuxiliaryProcessProxy object. The reason for this I did not want to pass an AuxiliaryProcessProxy
object that is in the middle of its destructor to another function.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _forceNetworkProcessToTaskSuspendForTesting]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::~AuxiliaryProcessProxy):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
(WebKit::AuxiliaryProcessProxy::platformStartConnectionTerminationWatchdog):
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAndUIAssertion::ProcessAndUIAssertion):
* Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.h: Removed.
* Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.mm: Removed.
* Source/WebKit/UIProcess/ProcessAssertion.cpp:
(WebKit::ProcessAssertion::acquireAssertion):
(WebKit::ProcessAssertion::create):
(WebKit::ProcessAndUIAssertion::create):
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAndUIAssertion::ProcessAndUIAssertion):
(WebKit::ProcessAssertion::aquireAssertion): Deleted.
* Source/WebKit/UIProcess/ProcessAssertion.h:
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NetworkProcess.mm:
(isTaskSuspended):
(TEST(NetworkProcess, TerminateWhenNetworkProcessIsSuspended)):
* Tools/TestWebKitAPI/Utilities.h:
(TestWebKitAPI::Util::waitFor):

Canonical link: <a href="https://commits.webkit.org/289075@main">https://commits.webkit.org/289075@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af966fad9fd92c72308a7ef4ebdca5a277725913

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85117 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4844 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90262 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36170 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87205 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4934 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12821 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66202 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24030 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88162 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3745 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77327 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46472 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3654 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31569 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35243 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74423 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32394 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91659 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12458 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9084 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74716 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12687 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73158 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73834 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18296 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18234 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16666 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4472 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12406 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17882 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12242 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15737 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13988 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -89,6 +89,9 @@ AuxiliaryProcessProxy::AuxiliaryProcessProxy(ShouldTakeUIBackgroundAssertion sho
 
 AuxiliaryProcessProxy::~AuxiliaryProcessProxy()
 {
+    if (state() != State::Terminated)
+        platformStartConnectionTerminationWatchdog();
+
```
I wonder if it's a viable option to call this in the NetworkProcessProxy destructor (and possibly the other proxy destructors), and avoid the ProcessAssertion changes, since the object will not be partially destroyed, then?

pvollan replied:
The reason I am wondering this is that I think it may make sense to keep having the `ProcessAssertion` constructor take an `AuxiliaryProcessProxy&` parameter, since we normally only want to take assertions on these process types.

bnham replied:
I did leave in the `ProcessAssertion::create` and `ProcessAndUIAssertion::create` variants that take in an `AuxiliaryProcessProxy` as an argument since that's probably the main way you want to interact with those classes.

But calling `ProcessAndUIAssertion::create(*this, ...)` from the NetworkProcessProxy destructor doesn't solve anything, since the problem with passing `this` associated with an object that's being destructed to another function is that the other function could inadvertently `ref()` it, which would cause a UAF: the pointer to the destructed object will be stored in that Ref variable, and the user of that Ref will think the object has a non-zero ref count, but it's actually already been (or will soon be) deleted. That doesn't actually happen the way the code is currently written, but I guess I am being pedantic and trying to avoid someone accidentally causing that to happen in the future.

It also violates the "safer C++" guidelines that there has to be a RefPtr protecting any ref-counted argument on the stack...but clearly there can't be a RefPtr protecting `this` on the stack if you're in the destructor, otherwise the destructor wouldn't have started running in the first place.

pvollan replied:
Ah, yes, that is a very good point, thanks!

pvollan reviewed (approved):
R=me.

bnham committed:
NetworkProcess does not always terminate properly on macOS
https://bugs.webkit.org/show_bug.cgi?id=285947
rdar://142838090

Reviewed by Per Arne Vollan.

NetworkProcess does not always terminate on macOS. The reason for this is that NetworkProcess can
suspend in the middle of termination:

1. The NetworkProcessProxy destructor drops "m_backgroundActivityToPreventSuspension", which allows
NetworkProcess to suspend.
2. Then the AuxiliaryProcessProxy destructor runs and starts termination by invalidating via
IPC::Connection::invalidate.
3. The remote side of the connection invalidation in NetworkProcess (NetworkProcess::didClose) might
never run if the process is already suspended, which then leaves the process in a permanently
suspended state.

To fix this, as part of AuxiliaryProcessProxy destruction, we now start a watchdog timer which
allows the process to keep running while termination is active, and we kill the process when the
timer expires. This used to be part of XPCConnectionTerminationWatchdog, but I removed that class
since it wasn't really doing much anymore after this refactoring.

The main refactoring was making it so that process assertions can be created without an
AuxiliaryProcessProxy object. The reason for this I did not want to pass an AuxiliaryProcessProxy
object that is in the middle of its destructor to another function.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _forceNetworkProcessToTaskSuspendForTesting]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::~AuxiliaryProcessProxy):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
(WebKit::AuxiliaryProcessProxy::platformStartConnectionTerminationWatchdog):
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAndUIAssertion::ProcessAndUIAssertion):
* Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.h: Removed.
* Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.mm: Removed.
* Source/WebKit/UIProcess/ProcessAssertion.cpp:
(WebKit::ProcessAssertion::acquireAssertion):
(WebKit::ProcessAssertion::create):
(WebKit::ProcessAndUIAssertion::create):
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAndUIAssertion::ProcessAndUIAssertion):
(WebKit::ProcessAssertion::aquireAssertion): Deleted.
* Source/WebKit/UIProcess/ProcessAssertion.h:
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NetworkProcess.mm:
(isTaskSuspended):
(TEST(NetworkProcess, TerminateWhenNetworkProcessIsSuspended)):
* Tools/TestWebKitAPI/Utilities.h:
(TestWebKitAPI::Util::waitFor):

Canonical link: https://commits.webkit.org/289075@main
